lib/spack/spack/fetch_strategy.py

```diff
@@ -1504,6 +1537,15 @@ def for_package_version(pkg, version):
     if not isinstance(version, Version):
         version = Version(version)
 
+    # if it's a commit, we must use a GitURLFetcher
+    if version.is_commit and hasattr(pkg, "git"):
+        # Populate the version with comparisons to other commits
+        repository = git_repo_for_package(pkg)
+        fetcher = GitFetchStrategy(git=repository)
+        version.generate_commit_lookup(fetcher, pkg.versions)
```

Comment:
All calls to `generate_commit_lookup` have to create a GitFetchStrategy -- I would just move the creation of the special fetch strategy into the `generate_commit_lookup` and pass the repository. I think that ends up being cleaner (it's really `generate_commit_lookup`'s fetcher).